modules/data.land/R/extract_soil_nc.R

```diff
@@ -113,61 +229,89 @@ extract_soil_gssurgo<-function(outdir, lat, lon, size=1, radius=500, depths=c(0.
+          if (any(is.na(DepthL.Data$soil_organic_carbon_stock))) {
+            PEcAn.logger::logger.warn("Found NA values in soil_organic_carbon_stock data. Removing incomplete records.")
+            DepthL.Data <- DepthL.Data[!is.na(DepthL.Data$soil_organic_carbon_stock), ]
+          }
+          if (nrow(DepthL.Data) == 0) {
+            PEcAn.logger::logger.warn("No valid SOC data after removing NAs")
+            return(NULL)
+          }
+          # Simulate SOC uncertainty using Gamma distribution
+          soc_mean <- mean(DepthL.Data$soil_organic_carbon_stock, na.rm = TRUE)
+          soc_sd <- stats::sd(DepthL.Data$soil_organic_carbon_stock, na.rm = TRUE)
+          
+          # Handle edge cases for SOC simulation
+          if (nrow(DepthL.Data) == 1) {
+            simulated_soc <- rep(NA_real_, size)
+          } else if (is.na(soc_sd) || soc_sd == 0) {
+            simulated_soc <- rep(NA_real_, size)
+          } else {
+            shape <- (soc_mean^2) / (soc_sd^2)
+            rate <- soc_mean / (soc_sd^2)
+            simulated_soc <- pmax(stats::rgamma(size, shape=shape, rate=rate), 0)
+          }
+          
           simulated.soil<-simulated.soil %>%
             as.data.frame %>%
-            dplyr::mutate(DepthL=rep(DepthL.Data[1,6], size),
-                   mukey=rep(DepthL.Data[1,5], size)) %>%
+            dplyr::mutate(DepthL=rep(DepthL.Data$DepthL[1], size),
+                   mukey=rep(DepthL.Data$mukey[1], size),
+                   soil_organic_carbon_stock = simulated_soc) %>%
             `colnames<-`(c("fraction_of_sand_in_soil",
                            "fraction_of_silt_in_soil",
                            "fraction_of_clay_in_soil",
                            "soil_depth",
-                           "mukey"))
+                           "mukey",
+                           "soil_organic_carbon_stock"))
           simulated.soil
         },
         error = function(e) {
           PEcAn.logger::logger.warn(conditionMessage(e))
           return(NULL)
         })
-        
       }) 
     
     # estimating the proportion of areas for those mukeys which are modeled
-    mukey_area <- mukey_area %>%
-      dplyr::filter(mukeys %in% simulated.soil.props$mukey) %>%
-      dplyr::mutate(Area=.data$Area/sum(.data$Area))
-    
+
+    # defining mukey_area
+    mukey_area <- data.frame(
+      mukey = names(mukey_counts),
+      Area = as.numeric(mukey_counts) / sum(mukey_counts)
+    ) %>%
+      dplyr::filter(mukey %in% unique(simulated.soil.props$mukey)) %>%
+      dplyr::mutate(Area = Area / sum(Area, na.rm = TRUE))
     #--- Mixing the depths
     soil.profiles<-simulated.soil.props %>% 
-      split(.$mukey)%>% 
+      split(.$mukey) %>%   
       purrr::map(function(soiltype.sim){
-        sizein <- (mukey_area$Area[ mukey_area$mukey == soiltype.sim$mukey %>% unique()])*size
+        sizein <- (mukey_area$Area[ mukey_area$mukey == unique(soiltype.sim$mukey)[1]])*size
```

Comment:
Both the `%in%` and `== unique()[1]` obscure the fact that the `split` on `mukey` enforces the fact that `unique(soiltype.sim$mukey)` will be a single value. I'd suggest:

```suggestion
        sizein <- mukey_area$Area[mukey_area$mukey == unique(soiltype.sim$mukey) * size
```

Reply:
thanks! done 👍

Reply:
done 👍